### PR TITLE
PYTHON-3454 Specifying a generic type for a collection does not correctly enforce type safety when inserting data

### PR DESCRIPTION
**Pull request by juliusgeo**

### CODE REVIEW

File: test/test_mypy.py [hunk 1]
```diff
@@ -18,13 +18,25 @@
 import os
 import tempfile
 import unittest
-from typing import TYPE_CHECKING, Any, Dict, Iterable, Iterator, List
+from typing import TYPE_CHECKING, Any, Dict, Iterable, Iterator, List, Optional
 
 try:
-    from typing import TypedDict  # type: ignore[attr-defined]
+    from typing_extensions import NotRequired, TypedDict
```
Comment (ShaneHarvey) [inline test/test_mypy.py:24]: Doesn't this mean these tests will be skipped even when TypedDict exists because typing_extensions is not installed in CI?

File: test/test_mypy.py [hunk 2]
```diff
@@ -304,6 +324,62 @@ def test_typeddict_document_type(self) -> None:
...
+    def test_typeddict_document_type_insertion(self) -> None:
+        client: MongoClient[Movie] = MongoClient()
+        coll: Collection[Movie] = client.test.test
+        mov = Movie(_id=ObjectId(), name="THX-1138", year=1971)
+        for meth, arg in [
+            (coll.insert_many, [mov]),
+            (coll.insert_one, mov),
+            (coll.bulk_write, [InsertOne(mov)]),
+        ]:
+            meth(arg)  # type:ignore[operator]
```
Comment (ShaneHarvey) [inline test/test_mypy.py:337]: How can we put a type ignore here? insert_many/insert_one/bulk_write working without raising a mypy error is the main purpose of this bug fix. I think we're loosing sight of what's important here because the PR is muddled by tests for find_one/_id/NotRequired. Can you open a new ticket for the NotRequired stuff? To fix the bug in PYTHON-3454 I believe we only need to mention schema validation should be configured on the server and add a test like this to confirm the bug is fixed:
```python
    def test_typeddict_document_type_insertion(self) -> None:
        coll: Collection[Movie] = self.db.test
        mov = Movie(name="THX-1138", year=1971)
        coll.insert_one(mov)
        coll.insert_many([mov])
        coll.bulk_write([InsertOne(mov)])
```

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]

File: test/test_mypy.py (hunk 2 above)
Comment (ShaneHarvey) [reply test/test_mypy.py]: Let's remove everything about _id, NotRequired, typing_extensions, and all the other tests besides the one above. All that is extra stuff that's only tangentially related to the issue. We already have tests that find_one correctly yields the collection generic type: https://github.com/mongodb/mongo-python-driver/blob/4.3.2/test/test_mypy.py#L299-L305

Comment (ShaneHarvey) [reply test/test_mypy.py]: There are two complaints in [PYTHON-3454](https://jira.mongodb.org/browse/PYTHON-3454):
1) `collection.insert_one(Person({"name": "Harold", "age": 15}))` incorrectly raises a mypy error. This is fixed by using _DocumentType and tested by `test_typeddict_document_type_insertion` above.
2) `collection.insert_one({"name": 17, "age": "this isn't a number", "gender": 6})` does not raise a mypy error when it should. This is also fixed by using  _DocumentType and should also be tested (if we aren't doing it already).

This is why I see the _id, NotRequired, typing_extensions, and find test as all unrelated. They are independent from the 2 issues reported by the user.

Comment (ShaneHarvey) [reply test/test_mypy.py]: > Perhaps I am misunderstanding, but I don't think that the user wanted to be able to insert a dict and have that be type-checked.

They do want this. In fact that's all a TypeDict is, it's a regular dict that's been type checked. So both of these inserts are completely identical and should have the same behavior:
```python
collection: Collection[Person] = client.t.t
collection.insert_one(Person({"name": "Harold", "age": 15})) 
collection.insert_one({"name": "Harold", "age": 15}) 
```
And both of these line should also cause the same mypy error:
```python
collection: Collection[Person] = client.t.t
collection.insert_one(Person({"name": "Harold", "age": "WRONG TYPE"})) 
collection.insert_one({"name": "Harold", "age": "WRONG TYPE"}) 
```
Both of these issues will be fixed by this PR. 

I agree that the subtleties of _id + TypedDict are worth documenting but I would like you to open a new jira ticket for it that we can prioritize separately. It's a different issue from PYTHON-3454.

Comment (blink1073) [reply test/test_mypy.py]: In the original example the `TypeDict` did not have an explicit `_id` field.

Comment (blink1073) [reply test/test_mypy.py]: Not including the `_id` is still a valid choice, if it is not relevant to your application.

**Review by blink1073 — APPROVED** [review]
LGTM!

File: test/test_mypy.py [hunk 3]
```diff
@@ -304,6 +327,28 @@ def test_typeddict_document_type(self) -> None:
...
+        coll.insert_many([mov])  # type:ignore[list-item]
+        coll.insert_many([movie])
+        bad_mov = {"name": "THX-1138", "year": "WRONG TYPE"}
+        bad_movie = Movie(name="THX-1138", year="WRONG TYPE")  # type:ignore[typeddict-item]
+        coll.insert_one(bad_mov)  # type:ignore[arg-type]
+        coll.insert_one({"name": "THX-1138", "year": "WRONG TYPE"})  # type:ignore[typeddict-item]
+        coll.insert_one(bad_movie)
+        coll.insert_many([bad_mov])  # type:ignore[list-item]
+        coll.insert_many(
+            [{"name": "THX-1138", "year": "WRONG TYPE"}]
```
Comment (blink1073) [inline test/test_mypy.py:348]: The auto-formatter got us here, you need to move the ignore comment

**Comment by blink1073** [comment]
There's also some unused imports causing the lint to fail.

**Review by blink1073 — APPROVED** [review]

File: pymongo/operations.py [hunk 4]
```diff
@@ -19,15 +19,15 @@
...
+from pymongo.typings import _CollationIn, _DocumentIn, _DocumentType, _Pipeline
 
 
 class InsertOne(object):
     """Represents an insert_one operation."""
 
     __slots__ = ("_doc",)
 
-    def __init__(self, document: _DocumentIn) -> None:
+    def __init__(self, document: Union[_DocumentIn, _DocumentType]) -> None:
```
Comment (ShaneHarvey) [inline pymongo/operations.py:30]: Can we defer this change to PYTHON-3493?

File: doc/examples/type_hints.rst [hunk 5]
```diff
@@ -92,11 +92,13 @@ Note that when using :class:`~bson.son.SON`, the key and value types must be giv
...
 
-You can use :py:class:`~typing.TypedDict` (Python 3.8+) when using a well-defined schema for the data in a :class:`~pymongo.collection.Collection`:
+You can use :py:class:`~typing_extensions.TypedDict` (Python 3.8+) when using a well-defined schema for the data in a
+:class:`~pymongo.collection.Collection`. Note that all `schema validation`_ for inserts and updates is done on the server.
+These methods automatically add an "_id" field.
 
 .. doctest::
 
-  >>> from typing import TypedDict
+  >>> from typing_extensions import TypedDict
```
Comment (ShaneHarvey) [inline doc/examples/type_hints.rst:101]: Let's remove typing_extensions.

File: .github/workflows/test-python.yml [hunk 6]
```diff
@@ -59,14 +59,15 @@ jobs:
           cache-dependency-path: 'setup.py'
       - name: Install dependencies
         run: |
-          python -m pip install -U pip mypy
+          python -m pip install -U pip mypy typing_extensions
```
Comment (ShaneHarvey) [inline .github/workflows/test-python.yml:62]: Let's remove typing_extensions.

File: test/test_mypy.py [hunk 7]
```diff
@@ -14,22 +14,21 @@
...
 import tempfile
 import unittest
 from typing import TYPE_CHECKING, Any, Dict, Iterable, Iterator, List
 
 try:
-    from typing import TypedDict  # type: ignore[attr-defined]
-
     # Not available in Python 3.7
+    from typing_extensions import TypedDict
+
```
Comment (ShaneHarvey) [inline test/test_mypy.py:25]: Let's revert these changes and go back to using typing.

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]

File: test/test_mypy.py (hunk 7 above)
Comment (blink1073) [reply test/test_mypy.py:25]: We will need to keep using `typing_extensions` in this file, or else split off a different file to run `mypy` on Python 3.7.  The reason it was working before is that we were inadvertently ignoring errors when running this file.

File: test/test_mypy.py [hunk 8]
```diff
@@ -14,22 +14,17 @@
 
 """Test that each file in mypy_fails/ actually fails mypy, and test some
 sample client code that uses PyMongo typings."""
-
 import os
 import tempfile
 import unittest
 from typing import TYPE_CHECKING, Any, Dict, Iterable, Iterator, List
 
-try:
```
Comment (ShaneHarvey) [inline test/test_mypy.py:23]: We need to keep this in a try/except so that `python test/test_mypy.py` works even when typing_extensions is not installed.

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]

File: test/test_mypy.py (hunk 7 above)
Comment (ShaneHarvey) [reply test/test_mypy.py:23]: > The reason it was working before is that we were inadvertently ignoring errors when running this file.

Can you explain? How was this fixed?

Comment (blink1073) [reply test/test_mypy.py:24]: We added a new invocation in GitHub Actions in this PR to run this file without disabling several error checks.

Comment (ShaneHarvey) [reply test/test_mypy.py:24]: I see thanks.

**Review by ShaneHarvey — APPROVED** [review]
